### ISSUE TITLE
Issues on heterogenous action and observation space

### ISSUE THREAD

**Issue by UniGen-X** [liruiw/HPT#32]
Hi,

This is a really excellent work, and I am currently extending it to my own research. I have three different robots whose configurations are:

Robot 1:
action dim: 14
observation images: main_image

Robot 2:
action dim: 16
observation images: main_image, left_wrist_image, right_wrist_image

Robot 3:
action dim: 14
observation images: main_image, left_wrist_image, right_wrist_image

I have already customized my own `convert_dataset_image` functions, however, it ran into shape miss-matching issue under `replay_buffer.add_episode` on line(470) `assert value.shape[1:] == arr.shape[1:], f"key {key} shape issue`, saying:

**key action shape issue
key image shape issue**

I was wondering if you code originally support robots with different action and observation space. Thanks!

**Comment by liruiw** [comment]
Hi, the different action and observation spaces are supported in the pre-training code. Essentially you would need a module dict where you create each key for each dataset. For fine-tuning, we were assuming a single fixed embodiment at one time.

**Comment by UniGen-X** [comment]
Thank you for the response! Could you please point out where the corresponding line in the pre-training code? Actually, which exactly pre-training code do you mean? 

I am working on 'run.py', which seems not initialize separate stems and heads for different robots. Thanks!

**Comment by UniGen-X** [comment]
Also, two follow-up questions:

1. I noticed that in the paper, you mentioned that "**we use one camera view of the scene with the pre-trained frozen ResNet18 image encode to compute the vision features.**". I am assuming the "[HPT-Base (With Language)](https://huggingface.co/liruiw/hpt-base-lang)" is pretrained with one camera view, right? If so, to fine-tuning the model with multiple-views would be problematic, right? 
2. any hint regarding how many epochs in general needed to fine-tuning on a new task? 

Thank you very much!

**Comment by liruiw** [comment]
Finetuning on multiple epochs won't be problematic. The general hints for fine-tuning is to train longer first and then reduce the training times if you see overfitting.

**Comment by UniGen-X** [comment]
Thank you for the response! What about another question above regarding the pre-training code that allows training over heterogenous robots? Is it already released? Thanks!

**Comment by liruiw** [comment]
I wasn't able to get around to release the code. Feel free to look at the HMA code https://github.com/liruiw/HMA which includes the pre-training and might have some details.
